File: oniguruma-ffm/src/test/java/me/zolotov/oniguruma/ffm/OnigurumaTest.java
package me.zolotov.oniguruma.ffm;

import org.junit.jupiter.api.Test;

import java.nio.charset.StandardCharsets;

import static org.junit.jupiter.api.Assertions.*;

class OnigurumaTest {
    @Test
    void invalidPatternRaisesRuntimeException() {
        try (var oniguruma = Oniguruma.createFromResources()) {
            var exception = assertThrows(
                    RuntimeException.class,
                    () -> oniguruma.createRegex("(".getBytes(StandardCharsets.UTF_8))
            );
            assertTrue(
                    exception.getMessage().toLowerCase().contains("parenthes")
                            || exception.getMessage().toLowerCase().contains("compile"),
                    () -> "unexpected error message: " + exception.getMessage()
            );
        }
    }

    @Test
    void matchRejectsByteOffsetBeyondContentLength() {
        try (var oniguruma = Oniguruma.createFromResources()) {
            var regex = oniguruma.createRegex("a".getBytes(StandardCharsets.UTF_8));
            var text = oniguruma.createString(new byte[0]);

            var exception = assertThrows(
                    IllegalArgumentException.class,
                    () -> oniguruma.match(regex, text, 1, true, true)
            );
            assertEquals("byteOffset 1 out of range [0, 0]", exception.getMessage());

            oniguruma.freeRegex(regex);
            oniguruma.freeString(text);
        }
    }
}

File: koniguruma/src/commonTest/kotlin/me/zolotov/oniguruma/OnigurumaTest.kt
package me.zolotov.oniguruma

import kotlinx.coroutines.test.TestResult
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

class OnigurumaTest {
    @Test
    fun invalidPatternRaisesOnigurumaException(): TestResult = runTest {
        createOniguruma().use { oniguruma ->
            val exception = assertFailsWith<OnigurumaException> {
                oniguruma.createRegex("(".encodeToByteArray())
            }
            // Assert on the oniguruma diagnostic itself, not just on our own wrapper text, so a
            // backend that swallows the detail fails the test.
            assertTrue(
                exception.message.orEmpty().lowercase().contains("parenthes"),
                "error message should carry the oniguruma diagnostic, got: ${exception.message}"
            )
        }
    }

    @Test
    fun matchRejectsByteOffsetBeyondContentLength(): TestResult = runTest {
        createOniguruma().use { oniguruma ->
            oniguruma.createRegex("a".encodeToByteArray()).use { regex ->
                oniguruma.createString(ByteArray(0)).use { text ->
                    val exception = assertFailsWith<IllegalArgumentException> {
                        oniguruma.match(regex, text, 1, matchBeginPosition = true, matchBeginString = true)
                    }
                    assertEquals("byteOffset 1 out of range [0, 0]", exception.message)
                }
            }
        }
    }
}
